### ISSUE TITLE
[FR] Tighter fzf integration - use fzf for all ble.sh menus

### ISSUE THREAD

**Issue by pallaswept** [akinomyoga/ble.sh#479]
As I begin to type this, I realise that I should begin by saying, thank you!

I have really been enjoying ble.sh. One feature I use a lot is the menus. As a disabled user I have been finding a greatly reduced load on my hands at the terminal, which comes from being able to pick things out of a list with cursor keys, which is much easier than typing, for me.

So, thanks @akinomyoga for unloading a disabled person's very sore hands!!!

I did find the initial experience a little jarring, though - through my own fault, not blesh's fault. See, I was using fzf before blesh, but only a little... so when I saw that blesh had fzf integration, I just turned it on. I didn't really know what I was doing, honestly. fzf is new to me (I've been away from linux a long time) and now I'm trying blesh which is also new, and then I put these two new things together mindlessly.

I would be using the terminal, and try something like hitting tab during a command, and I would get a cool menu that was awesome. But then sometimes, I would get a different menu that was also awesome but it was different. I didn't know what was happening at first, but I quickly realised - I was seeing fzf menus sometimes - which I thought were blesh... and then sometimes I'd see blesh's menus.

I just 'got used to it' and there is no real problem here. One menu is not always 'better' than the other type of menu... but over time I grew to think - Perhaps it could be cool if fzf were used to render all the ble.sh menus? That way, the user would have a nice consistent experience all the time. 

It seems like fzf could handle this quite well, too, for example its 'preview' window could be leveraged to show descriptions when auto-complete the options of a command. I do know that blesh has lots of options for the menu completion, so perhaps fzf could not support all of them, but many would be adaptable, so perhaps this idea could still work?

Thanks again for your time.

**Comment by akinomyoga** [comment]
Something like this? Please adjust the behavior of the selector (the shell function `belrc/selector`) depending on your preferences.

```bash
# blerc

## @fn blerc/selector args...
##   Returns the index of the selected candidate.
##   @param[in] args...
##     The list of candidates. Each argument takes the form "<index> <word> ...".
##   @stdout
##     Output the selected index.
function blerc/selector {
  printf '%s\n' "$@" | fzf --with-nth=2.. | cut -d ' ' -f 1
}

function blerc/select-and-insert {
  # Create a list of "<index> <word> - <desc>" where ^\ is the control character \x1c
  local -a list=()
  local i
  for ((i=0;i<cand_count;i++)); do
    local ACTION=${cand_pack[i]%%:*} desc=
    ble/is-function ble/complete/action:"$ACTION"/get-desc &&
      local "${_ble_complete_cand_varnames[@]/%/=}" &&
      ble/complete/cand/unpack "${cand_pack[i]}" &&
      ble/complete/action:"$ACTION"/get-desc
    (($?==148)) && return 148

    ble/array#push list "$i ${cand_cand[i]} - $desc"
  done

  ble/term/leave-for-widget
  local index=$(blerc/selector "${list[@]}")
  ble/term/enter-for-widget
  ble/textarea#invalidate

  if [[ $index ]]; then
    cand_count=1
    cand_cand=("${cand_cand[index]}")
    cand_word=("${cand_word[index]}")
    cand_pack=("${cand_pack[index]}")
    ble/complete/insert-common
  fi
  return 148
}
ble-import -C 'ble/function#push ble/complete/menu/show "blerc/select-and-insert"' core-complete
```

**Comment by pallaswept** [comment]
Whoa that's amazing. 

I thought that maybe it might be possible to override an internal ble.sh function for this. I did NOT think it was possible to do it so fast. I feared I might be starting a months-long project, you did it in a few hours :D Amazing.

Thank you!!

**Comment by akinomyoga** [comment]
@pallaswept I added further adjustments to the above example and added it at [`contrib/integration/fzf-menu`](https://github.com/akinomyoga/blesh-contrib/blob/master/integration/fzf-menu.bash). The adjustments include the colors of the items and the alignment of the descriptions. After updating ble.sh by `ble-update`, one can now use it with the following setting:

```bash
# ~/.blerc

ble-import -d integration/fzf-menu
```

**Comment by pallaswept** [comment]
This is so cool!!

I found one tiny bug, I thought I should mention it here before I do a CR, but I can file one if you like? It's this:

https://github.com/akinomyoga/blesh-contrib/blob/b34bec2b8eae89c412fb65ffab8fd21412a3f5c8/integration/fzf-menu.bash#L48
```
  builtin eval -- "$formatter" | fzf "${fzf_options[@]}" | cut -d ' ' -f 1
```

I got some bugs with this, but this seems to help:
```
  builtin eval -- "$formatter" | fzf "${fzf_options[@]}" | cut -d $sep -f 1
```
I just change the cut on `' '` to a cut on `$sep`. This ensures that the correct index is returned. I was having trouble where for example if I choose item `8` on the list called `FooBar`, the Index returned was `8FooBar` rather than just `8`

Here is an example of such an error. This is seen after pressing enter to select the item in the fzf menu. The item was "Documents" and the index was 36:
```
> cd bash: 36Documents: syntax error: invalid arithmetic operator (error token is "Documents")
[ble: press RET to continue]
```

A very minor thing, not even a bug, just a typo: "inetgration" appears in place of "integration" [1](https://github.com/akinomyoga/blesh-contrib/blob/b34bec2b8eae89c412fb65ffab8fd21412a3f5c8/integration/fzf-menu.bash#L58) [2](https://github.com/akinomyoga/blesh-contrib/blob/b34bec2b8eae89c412fb65ffab8fd21412a3f5c8/integration/fzf-menu.bash#L109) [3](https://github.com/akinomyoga/blesh-contrib/blob/b34bec2b8eae89c412fb65ffab8fd21412a3f5c8/integration/fzf-menu.bash#L119)

I hope this is helpful! I'm loving these recent updates, very exciting :)
